### ISSUE TITLE
Seg Fault for case involving symbolic lists through llvm backend

### ISSUE THREAD

**Issue by anutosh491** [lcompilers/lpython#2450]
Consider the following example 
```
from lpython import S
from sympy import Symbol, pi

def mmrv() -> list[S]:
    l1: list[S] = [pi]
    return l1

def test_mrv1():
    ans: list[S] = mmrv()
    el: S = ans[0]
    print(el)

test_mrv1()
```
We get
```
(lf) anutosh491@spbhat68:~/lpython/lpython$ lpython --enable-symengine --backend=llvm examples/expr2.py 
Segmentation fault
```

The same function would work for a list of any other type but fails to work for a symbolic list.

**Comment by anutosh491** [comment]
After debugging this I see that the error occurs due to the extraction of element from the list `ans[0]`.
The final ASR after all passes look good to me.
While going through the llvm IR I see 
```
  %26 = load void*, void** %el, align 8
  ......
  %31 = getelementptr %list, %list* %ans, i32 0, i32 2
  %32 = load void**, void*** %31, align 8
  %33 = getelementptr inbounds void*, void** %32, i32 0
  %34 = load void*, void** %33, align 8
  call void @basic_assign(void* %26, void* %34)
  %35 = load void*, void** %el, align 8
```

Here I think the `basic_assign` assigns `%34` to `%26` but then it is not being stored back into `%el` and hence when line 
`%35 = load void*, void** %el, align 8` is executed, `%el` isn't having the required value. I think we might want to have a 
`store void** %26, void*** %el, align 8`

**Comment by certik** [comment]
This works:
```
def mmrv() -> list[i32]:
    l1: list[i32] = [5]
    return l1

def test_mrv1():
    ans: list[i32] = mmrv()
    el: i32 = ans[0]
    print(el)

test_mrv1()
```
So we can compare it against the CPtr version after the symbolic pass and see what the problem is. You can then try to reproduce it by explicitly writing Python code that uses CPtr (does not use the symbolic pass) and we need to figure out why it segfaults.

**Comment by anutosh491** [comment]
I was almost able to figure this out. Now that #2453 & #2454 have gone in, we have the llvm and c backend ready to deal with any symbolic container like a symbolic list.

So consider the following. We expect the program given above to transform to something like this (in python) after the **symbolic** and the **subroutine_from_function** pass
```
from lpython import S
from sympy import Symbol, pi

def mmrv(r: Out[list[S]]) -> None:
    x: S = pi
    l1: list[S] = [x]
    r = l1

def test_mrv2():
    ans: list[S]
    temp: list[S]
    mmrv(temp)
    ans = temp
    print(ans[0])

test_mrv2()
```
Now this gives us a segmentation fault both through the llvm and the c backend 
```
(lf) anutosh491@spbhat68:~/lpython/lpython$ lpython --enable-symengine integration_tests/test_gruntz.py --backend=c
Segmentation fault
(lf) anutosh491@spbhat68:~/lpython/lpython$ lpython --enable-symengine integration_tests/test_gruntz.py --backend=llvm
Segmentation fault
```
But what works is this 
```
from lpython import S
from sympy import Symbol, pi

def mmrv(r: Out[list[S]]) -> None:
    x: S = pi
    l1: list[S] = [x]
    r = l1

def test_mrv2():
    ans: list[S]
    mmrv(ans)
    print(ans[0])

test_mrv2()

(lf) anutosh491@spbhat68:~/lpython/lpython$ lpython --enable-symengine integration_tests/test_gruntz.py --backend=c
pi
(lf) anutosh491@spbhat68:~/lpython/lpython$ lpython --enable-symengine integration_tests/test_gruntz.py --backend=llvm
pi
```

**Comment by anutosh491** [comment]
So basically the **subroutine_from_function** pass, alters two steps which are 
1) passing a `temp` variable (`__libasr__created__var__0__func_call_res`) to the SubRoutine
2) Assign the `temp` variable back to `ans`

Not that this is wrong but the step 2 here is responsible for the issues through both backends. Considering the C backend. All code is same for both the code snippets above except 3 lines 
```
void test_mrv2()
{
    struct list_CPtr ans;
    mmrv(&ans);
    printf("%s\n", basic_str(ans.data[0]));
}

void test_mrv2()
{
    struct list_CPtr ans;
    struct list_CPtr temp;
    mmrv(&temp);
    list_deepcopy_CPtr(&temp, &ans);

    printf("%s\n", basic_str(ans.data[0]));
}
```
The first one works the second gives us a segmentation fault.

**Comment by Thirumalai-Shaktivel** [comment]
Here, the following doesn't seg fault. I think the issue might be either with print statement or `basic_str`. What do you think?
```py
from lpython import S
from sympy import Symbol, pi

def mmrv(r: Out[list[S]]) -> None:
    x: S = pi
    l1: list[S] = [x]
    r = l1

def test_mrv2():
    ans: list[S]
    temp: list[S]
    mmrv(temp)
    ans = temp
    x: S = ans[0]
    print(x)

test_mrv2()
```

**Comment by anutosh491** [comment]
Wait it doesn't seg fault for you ? It does for me !

**Comment by certik** [comment]
On the last Python example, both of these things fail for me:
```console
$ lpython --backend=c a.py -o xx --enable-symengine
$ ./xx                                             
zsh: bus error  ./xx
$ lpython --backend=llvm a.py -o yy --enable-symengine
$ ./yy 
zsh: segmentation fault  ./yy
```

We need to debug it and fix it.

**Comment by anutosh491** [comment]
@certik @Thirumalai-Shaktivel 
I have been trying to debug this but sadly haven't got any success yet.

I am not sure but I think I am stuck with a couple of issues. 
1) Consider the following
```
 from lpython import S
from sympy import Symbol, pi

def mmrv(r: Out[list[S]]) -> None:
    x: S = pi
    l1: list[S] = [x]
    r = l1

def test_mrv2():
    ans: list[S]
    mmrv(ans)
    x: S = ans[0]
    print(x)

test_mrv2()
```
When I execute it through lpython, it works
```
(lf) anutosh491@spbhat68:~/lpython/lpython$ lpython --enable-symengine --backend=c integration_tests/test_gruntz.py 
pi
```
But when I copy the generated C code (from `test_gruntz_tmp__generated__.c` to `expr.c` )
```
(lf) anutosh491@spbhat68:~/lpython/lpython$ clang -I${CONDA_PREFIX}/include -c expr.c -o expr.o -I src/libasr/runtime/
(lf) anutosh491@spbhat68:~/lpython/lpython$ clang -o expr.out expr.o -L$CONDA_PREFIX/lib -Wl,-rpath -Wl,$CONDA_PREFIX/lib -lsymengine -L src/runtime -llpython_runtime
(lf) anutosh491@spbhat68:~/lpython/lpython$ LD_LIBRARY_PATH=src/runtime ./expr.out
Segmentation fault
```
Not sure why this might be happening. I think the commands I'm executing are correct

**Comment by anutosh491** [comment]
2) I thought this has something to do with `list_deepcopy_CPtr` but then I checked it out through multiple different scenarios and I think what might be at fault is the `basic_str` function. Consider the following C code 
```
void mmrv(struct list_CPtr *r)
{
    int64_t _x;
    struct list_CPtr l1;
    void* x;
    _x = 0;
    x = NULL;
    x = (void*) &_x;
    basic_new_stack(x);
    basic_const_pi(x);
    struct list_CPtr constname0;
    list_init_CPtr(&constname0, 1);
    constname0.data[0] = x;
    constname0.current_end_point = 1;
    list_deepcopy_CPtr(&constname0, &l1);

    list_deepcopy_CPtr(&l1, r);
    printf("%p\n", r->data[0]);
    printf("%s\n", basic_str(r->data[0]));

    basic_free_stack(x);
}

void test_mrv2()
{
    int64_t _x;
    struct list_CPtr ans;
    struct list_CPtr temp;
    void* x;
    _x = 0;
    x = NULL;
    x = (void*) &_x;
    basic_new_stack(x);
    mmrv(&temp);

    // Print information about temp
    printf("temp capacity: %d\n", temp.capacity);
    printf("temp current_end_point: %d\n", temp.current_end_point);
    printf("temp data:\n");
    for (int i = 0; i < temp.current_end_point; ++i) {
        printf("  %p\n", temp.data[i]);
        printf("%s\n", basic_str(temp.data[i]));
    }

    list_deepcopy_CPtr(&temp, &ans);

    basic_assign(x, ans.data[0]);
    printf("%s\n", basic_str(x));
    basic_free_stack(x);
}
```
We are trying to print the exact same things from both the functions `mmrv` and `test_mrv2`. Even the **reference** being printed out is the same. And the result I get is this
```
(lf) anutosh491@spbhat68:~/lpython/lpython$ LD_LIBRARY_PATH=src/runtime ./expr.out
0x7ffc88087b00
pi
temp capacity: 1
temp current_end_point: 1
temp data:
  0x7ffc88087b00
Segmentation fault
```

**Comment by anutosh491** [comment]
Can be closed as the issue has been correctly fixed through #2465 & #2477
